### DISCUSSION TITLE
New placement - 'cover'

### DISCUSSION THREAD

**Discussion by KutnerUri** [floating-ui/floating-ui#1233] [Ideas]
## Feature description

I used Popper to create a highlight effect, using a div with border and `pointer-events: none;`.
I think this could be a cool feature of Popper (or an official modifier), as it is uses a lot internals.

![image](https://user-images.githubusercontent.com/5400361/106747064-14e3ce00-662c-11eb-99e8-a4cc473b8745.png)

These are the 4 original modifiers I used to implement the feature:

```tsx
const popperModifiers = [
	ignorePopperSize,
	resizeToMatchReference,
	disableFlip,
	calcCoverCoordinates
];

// when calculating pop position, do not include popper-size
const ignorePopperSize: Modifier<string, {}> = {
	name: 'ignorePopperSize',
	enabled: true,
	phase: 'beforeRead',
	fn({ state }) {
		state.rects.popper = {
			...state.rects.popper,
			height: 0,
			width: 0,
		};
	},
}

export type ResizeToMatchReferenceOptions = {
	offset: [number, number];
};
const resizeToMatchReference: Modifier<string, ResizeToMatchReferenceOptions> = {
	name: 'resizeToMatchReference',
	enabled: true,
	phase: 'main',
	fn({ state, options }) {
		const { offset: [offsetX, offsetY] = [0, 0] } = options;
		const { width, height } = state.rects.reference;

		state.styles.popper = {
			...state.styles.popper,
			width: width + offsetX + 'px',
			height: height + offsetY + 'px',
		};
	},
};

const disableFlip = 	{
	name: 'flip',
	enabled: false,
};

// regular popper pops away from target, shift popper down to the target, + constants offest
const calcCoverCoordinates = {
	name: 'coverCoordinates',
	options: {
		// move box from above the target ('top-start')
		// to make sure border covers the target from the outside.
		offset: ({ reference }: any) => [BASE_OFFSET, BASE_OFFSET - reference.height],
	},
},
```

### Offset
In hindsight, there are two of sizes consumers will like to control:
* top/left shift (move the entire popper vertically or horizontally)
* "padding" (increase/decrease size of the popper in each of the directions)

though, If I had to be compatible with the existing offset props, I'd go with
```
offset={[ ±aspectRatioPx, ±distanceFromCenterPx ]}
```

What do you think? I'm open for anything

**Comment by FezVrasta** [comment]
That's so cool! Thanks for posting it. I think the best way to share this would be to publish your own npm package, there are some guidelines on our documentation: https://popper.js.org/docs/v2/modifiers/community-modifiers/

**Reply by KutnerUri** [reply]
Thanks! burnt a few hours finding the right combination of modifiers.

Actually, I did publish them:
https://www.npmjs.com/package/@teambit/base-ui.utils.popper-js.ignore-popper-size
https://www.npmjs.com/package/@teambit/base-ui.utils.popper-js.resize-to-match-reference

I will try to tinker with it a little more and make it a single modifier / package, and respect the offset prop.
The package today is much easier to work with. :)

What bothers me is that I need to "undo" some of the existing logic.
For example, I need to undo Popper's trying to position the element above the target.
```
offset: ({ reference }) => [BASE_OFFSET, BASE_OFFSET - reference.height],
```
I would rather have a more stable way to do so.

**Reply by FezVrasta** [reply]
If you can think of some changes to apply to popper core to make this cleaner please send a PR.

Could you add the right keyword to the packages you published please? (`popperjs-modifier`)

**Reply by KutnerUri** [reply]
> Could you add the right keyword ?

I will try, my authoring tool [bit](https://bit.dev/) might not have this feature yet.

> If you can think of some changes to apply to popper core to make this cleaner please send a PR.

How about `placement="none"`?
In this case Popper would not **set** the position of the popper, and I will be able to calculate it with a modifier (instead of building a custom modifier chain, which will be less reusable).

In this case, I won't need `ignorePopperSize` and using the `-reference.height` at calcCoverCoordinates

**Reply by KutnerUri** [reply]
I'm willing to do the PR :)

**Reply by FezVrasta** [reply]
Sounds good!
